### ISSUE TITLE
SqlClient ManualTest MARSSyncTimeoutTest fails in managed mode

### ISSUE THREAD

**Issue by Wraith2** [dotnet/SqlClient#108]
When using the managed sni implementation in SqlClient the MARS related test MARSSyncTimeoutTest fails with an assertion in debug mode. This may not cause a problem in release builds because there are sufficient checks in the method body to prevent errors occurring or bubbling up to the user. I suspect that the assertion is incorrect but it needs scrutiny from someone more familiar with the library.

To reproduce on windows. Checkout master, change [UseManagedSNI](https://github.com/dotnet/corefx/blob/d4b014c7483789d3832d3c15fb3741e2cc5306b1/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObjectFactory.Windows.cs#L22) to true. Build and run manual tests, or run the specific named test If you know the commandline incantation required.

```
  Assertion Failed
  AsyncResult null on callback

     at System.Data.SqlClient.TdsParserStateObject.ReadAsyncCallback[T](IntPtr key, T packet, UInt32 error) in E:\Programming\csharp7\corefx\src\System.Data.SqlClient\src\System\Data\SqlClient\TdsParserStateObject.cs:line 2736
     at System.Data.SqlClient.TdsParserStateObject.ReadAsyncCallback[T](T packet, UInt32 error) in E:\Programming\csharp7\corefx\src\System.Data.SqlClient\src\System\Data\SqlClient\TdsParserStateObject.cs:line 2688
     at System.Data.SqlClient.SNI.SNIMarsHandle.HandleReceiveError(SNIPacket packet) in E:\Programming\csharp7\corefx\src\System.Data.SqlClient\src\System\Data\SqlClient\SNI\SNIMarsHandle.cs:line 321
     at System.Data.SqlClient.SNI.SNIMarsConnection.HandleReceiveError(SNIPacket packet) in E:\Programming\csharp7\corefx\src\System.Data.SqlClient\src\System\Data\SqlClient\SNI\SNIMarsConnection.cs:line 136
     at System.Data.SqlClient.SNI.SNIMarsConnection.HandleReceiveComplete(SNIPacket packet, UInt32 sniErrorCode) in E:\Programming\csharp7\corefx\src\System.Data.SqlClient\src\System\Data\SqlClient\SNI\SNIMarsConnection.cs:line 165
     at System.Data.SqlClient.SNI.SNIPacket.<>c__DisplayClass32_0.<ReadFromStreamAsync>b__0(Task`1 t) in E:\Programming\csharp7\corefx\src\System.Data.SqlClient\src\System\Data\SqlClient\SNI\SNIPacket.cs:line 276
     at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state) in E:\A\_work\104\s\src\System.Private.CoreLib\shared\System\Threading\ExecutionContext.cs:line 321
     at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread) in E:\A\_work\104\s\src\System.Private.CoreLib\src\System\Threading\Tasks\Task.cs:line 2448
     at System.Threading.ThreadPoolWorkQueue.Dispatch() in E:\A\_work\104\s\src\System.Private.CoreLib\src\System\Threading\ThreadPool.cs:line 621
    Finished:    System.Data.SqlClient.ManualTesting.Tests
```

It is happens when the socket for the connection is closed by the timeout in the test, in this case there is no data array in the `SNIPacket` and that means that the [assertion](https://github.com/dotnet/corefx/blob/d4b014c7483789d3832d3c15fb3741e2cc5306b1/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs#L2722) on [managed CheckPacket](https://github.com/dotnet/corefx/blob/d4b014c7483789d3832d3c15fb3741e2cc5306b1/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObjectManaged.cs#L38)  returns false indicating the packet is invalid by asking the packet if it's invalid, which checks the data. The [native CheckPacket](https://github.com/dotnet/corefx/blob/d4b014c7483789d3832d3c15fb3741e2cc5306b1/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObjectNative.cs#L102) doesn't check the contents of the packet, only that it is a pointer to a valid packet structure which is a slightly different check to the managed version, the managed version requires that the packet be non-null and that it contain a valid array which is the cause of the problem.

/cc @AfsanehR @saurabh500 @keeratsingh

**Comment by David-Engel** [comment]
@Wraith2 If you have time to look at this, great. Otherwise we will work on it when we have bandwidth. Thanks!

**Comment by Wraith2** [comment]
I posted a couple of solutions and discussed it at some length with @saurabh500 in their discussions. This is something we've discussed offline and it's waiting for him to get chance to investigate a fix which fits better with the architecture. 

https://github.com/dotnet/corefx/pull/37068#discussion_r277477889 
> Let me do a bit more digging around this area to provide more pointers and have a concrete explanation. Don't burn your cycles on this PR.

This is the main blocker on continued improvements to the managed implementation because it prevents many important tests from running.
